src/Illuminate/Database/Eloquent/Collection.php

```diff
@@ -600,4 +488,259 @@ public function getQueueableConnection()
 
         return $connection;
     }
+
+    /**
+     * We can not be sure the result of the following methods only holds models.
+     */
+
+    /**
+     * Convert to base collection if not solely filled with models.
+     *
+     * @param  static  $maybeEloquent
+     * @return \Illuminate\Support\Collection|static
+     */
+    protected static function maybeToBase($maybeEloquent)
+    {
+        return $maybeEloquent->contains(function ($item) {
```

Comment:
I think EloquentCollections assume all contained models to be the same class, so we might change this check to the following:

```php
        $model = Arr::first($maybeEloquent, Model::class);

        foreach($maybeEloquent as $maybeModel) {
            if(! $maybeEloquent instanceof $model) {
                return $maybeEloquent->toBase();
            }
        }

        return $maybeEloquent;
```

What do you think?